insights/parsers/mdstat.py

```diff
@@ -154,107 +157,118 @@ def parse_content(self, content):
 
     return personalities
 
 
 def parse_array_start(md_line):
-    """Parse the initial line of a device array stanza in
+    """
+    Parse the initial line of a device array stanza in
     ``/proc/mdstat``.
 
-    Lines are expected to be like:
+    Sample of md_line:
 
         md2 : active raid1 sdb3[1] sda3[0]
 
-    If they do not have this format, an error will be raised since it
-    would be considered an unexpected parsing error.
-
-    Parameters
-    ----------
-
-    md_line : str
-        A single line from the start of a device array stanza from a
-        ``/proc/mdstat`` file.
+    Args:
+        md_line:
+            A single line from the start of a device array stanza from a
+            ``/proc/mdstat`` file.
 
-    Returns
-    -------
-        A list of dictionaries, one dictionrary for each component
+    Returns:
+        A list of dictionaries, one dictionary for each component
         device making up the array.
+
+    Raises:
+        ParseException: If the format isn't like the sample above.
     """
+    auto_read_only = False
     components = []
+    device_flag = None
+    raid = None
+
+    # Split the line to create tokens, and
+    # set device_name to the first token.
     tokens = md_line.split()
     device_name = tokens.pop(0)
-    assert device_name.startswith("md")
-    assert tokens.pop(0) == ":"
+
+    if not device_name.startswith("md") or ":" not in tokens:
+        raise ParseException("The md line isn't as expected.")
+
+    # Remove the : symbol.
+    tokens.pop(0)
 
     active_string = tokens.pop(0)
-    active = False
     if active_string == "active":
         active = True
+    elif active_string == "inactive":
+        active = False
     else:
-        assert active_string == "inactive"
-
-    raid_read_only_token = tokens.pop(0)
-    auto_read_only = False
-    raid = None
-    if raid_read_only_token == "(auto-read-only)":
-        auto_read_only = True
-        raid = tokens.pop(0)
-    elif active:
-        raid = raid_read_only_token
-    else:  # Inactive devices don't list the raid type
-        raid = None
-        tokens.insert(0, raid_read_only_token)
+        raise ParseException("The raid isn't marked as active or inactive.")
+
+    # Only active raids have the auto-read-only
+    # entry or the raid level.
+    if active:
+        raid_read_only_token = tokens.pop(0)
+        if raid_read_only_token == "(auto-read-only)":
+            auto_read_only = True
+            raid = tokens.pop(0)
+        else:
+            raid = raid_read_only_token
 
     for token in tokens:
-        subtokens = re.split(r"\[|]", token)
-        assert len(subtokens) > 1
-        comp_name = subtokens[0]
-        assert comp_name
+        # Token should be sda1[0] or sda1[0](S) for
+        # example, and subtokens should be
+        # ['sda1', '0'] or ['sda1', '0', '(S)'].
+        subtokens = re.split(r"[\[\]]", token)
```

Comment:
Is there a neg/pos test that covers subtokens, I wasn't able to find it.

Reply:
It's in the test_parse_array_start function in the MD_FAIL test from what I can tell, let me know if I interrupted that test incorrectly.